pkg/actor/resize_pvc.go

```diff
@@ -0,0 +1,248 @@
+/*
+Copyright 2020 The Cockroach Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package actor
+
+import (
+	"context"
+	"fmt"
+	"github.com/cockroachdb/cockroach-operator/pkg/features"
+	"github.com/cockroachdb/cockroach-operator/pkg/kube"
+	"github.com/cockroachdb/cockroach-operator/pkg/utilfeature"
+	"time"
+
+	api "github.com/cockroachdb/cockroach-operator/api/v1alpha1"
+	"github.com/cockroachdb/cockroach-operator/pkg/condition"
+	"github.com/cockroachdb/cockroach-operator/pkg/resource"
+	"github.com/cockroachdb/cockroach-operator/pkg/update"
+	"github.com/cockroachdb/errors"
+	appsv1 "k8s.io/api/apps/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	kubetypes "k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// newResizePVC creates and returns a new resizePVC struct
+func newResizePVC(scheme *runtime.Scheme, cl client.Client, config *rest.Config) Actor {
+	return &resizePVC{
+		action: newAction("resize_pvc", scheme, cl),
+		config: config,
+	}
+}
+
+// resizePVC resizes a PVC
+type resizePVC struct {
+	action
+
+	config *rest.Config
+}
+
+// Handles returns true if the DB is initialized
+func (rp *resizePVC) Handles(conds []api.ClusterCondition) bool {
+	return condition.False(api.NotInitializedCondition, conds) && utilfeature.DefaultMutableFeatureGate.Enabled(features.ResizePVC)
+}
+
+// Act in this implementation resizes PVC volumes of a CR sts.
+func (rp *resizePVC) Act(ctx context.Context, cluster *resource.Cluster) error {
+	log := rp.log.WithValues("CrdbCluster", cluster.ObjectKey())
+
+	// If we do not have a volume claim we do not have PVCs
+	if cluster.Spec().DataStore.VolumeClaim == nil {
+		log.Info("Skipping PVC resize as VolumeClaim does not exist")
+		return nil
+	}
+
+	// Get the sts and compare the sts size to the size in the CR
+	key := kubetypes.NamespacedName{
+		Namespace: cluster.Namespace(),
+		Name:      cluster.StatefulSetName(),
+	}
+	statefulSet := &appsv1.StatefulSet{}
+	if err := rp.client.Get(ctx, key, statefulSet); err != nil {
+		return errors.Wrap(err, "failed to fetch statefulset")
+	}
+
+	// TODO statefulSetIsUpdating is not quite working as expected.
+	// I had to check status.  We should look at the update code in partition update to address this
+	if statefulSetIsUpdating(statefulSet) {
```

Comment:
Would this be better/possible to check in `Handles`?

Reply:
Yep.  I am planning on moving it.

Reply:
Looking at the code, I am not certain we can move it from here.  We need to get status working to help with this.